### ISSUE TITLE
USA Avenger weapon does not benefit from veterancy

### ISSUE THREAD

**Issue by ThePredatorBG** [TheSuperHackers/GeneralsGamePatch#2327]
Avenger's weapon does not benefit from veterancy because the turret is a PORTABLE_STRUCTURE. Veterancy bonus does not apply to separate turrets because they are different objects and Avenger experience does not transfer to the turret object.
If we can make separate turrets gain experience for themselves by killing enemies, that would make the damage bonuses work. Whether that's possible is not yet known.

Test map:
[TestAvengerVeterancy.zip](https://github.com/TheSuperHackers/GeneralsGamePatch/files/12536676/TestAvengerVeterancy.zip)

**Comment by commy2** [comment]
I presume WeaponBonusPassedToPassengers does not work with VETERAN, ELITE, HEROIC. Right?

**Comment by ThePredatorBG** [comment]
> I presume WeaponBonusPassedToPassengers does not work with VETERAN, ELITE, HEROIC. Right?

That actually worked.

**Comment by xezon** [comment]
Should the Avenger weapon become stronger when promoted? Would be quite a nice buff for Laser General. For all USA's against Airforce.

**Comment by commy2** [comment]
> That actually worked.

LOL

That also makes Frenzy and Propaganda work with Avenger then.

**Comment by commy2** [comment]
Should also double check Search and Destroy bonus.

**Comment by ReLaX82** [comment]
Well, I like this idea about boosting it.

Note: Check needs to be done acording to migs and buggies late game exspecially.

**Comment by Float1ngFree** [comment]
This is actually very tricky and has a lot more implications to consider.

Avengers have 100 instant dps (well, other AA nominally are  instant as well) which currently doesn't scale.
Quads - 50 dps and it scales with vet, upgrades & any effects you can capture.
Gattlig - 90 dps on max speed and up to 168+ just with Chain Guns + subliminal, not accounting for veterancy.

Also, please consider the possible +20% from bombardment. As well as attack speed increase.

Therefore, I'd say that if this change is to be approved it also needs to slightly reduce Avenger dps to accommodate bonuses from battle plans and further veterancy.

I've reduces the damage for both AA Lazer weapons from 10 to 9. 10% reduction even just for possible 20% gain from bombardment alone.

**Comment by xezon** [comment]
> Note: Check needs to be done acording to migs and buggies late game exspecially.

Aircraft, yes, but buggies? I think PointDefenseLaser is not a weapon and will not be affected by `WeaponBonusPassedToPassengers`.

> Also, please consider ... attack speed increase.

Attack speed increase how? Propaganda?

Avenger costs 1500 to 2000 to build. Gattling Tank and Quad Cannon cost 700 to 800. On paper, large damage output is justified. Many players just spam Humvee with TOW missile as their Anti Air, because the Avenger is not a great performer againt air units, if it was not for its ability to zip missiles and protect itself and nearby forces from missiles.

**Comment by Float1ngFree** [comment]
> Attack speed increase how? Propaganda?

Veterancy Rate of Fire increase, +20% per level.

Vet3 avenger with ditch 1.6 RoF * 1.3(1.5) Dmg (+Bombardment) = 2.08(2.4) times more dps than before patch. Or higher attack range than gat/quad with +20% from SnD. That's a thing to consider.

Unpopularity of Avenger IMO is mostly due to it slow speeds and therefore it fails to keep up with faster Humvee/Ambulance spam. With humvee mistakenly using a patriot missile - it is basically guaranteed to hit, so why bother with slow avenger?

Otherwise, it is fine as anti air, but lacks as a support unit because of how ridiculously OP missile infantry attack speed is and buggy barrage. Just too many missile objects. Laser designator could also use a buff in a form of change of stats it provides, 150% Rate of Fire is not bad per se, but I personally don't like the ridiculous attack speeds that are possible in original game.

**Comment by xezon** [comment]
Avenger only gains veterancy by killing aircraft, whereas Gattling Tank can kill ground units and Quad Cannon can comfortably kill ground units, buildings and collect scrap for veterancy.

Also, Avenger requires twice as many experience to get promoted, which is negated by the Advanced Training upgrade.

```
Object AmericaTankAvenger
  ExperienceRequired = 0 200 300 600

Object GLAVehicleQuadCannon
  ExperienceRequired = 0 100 150 300

Object ChinaTankGattling
  ExperienceRequired = 0 80 200 300
```

But in practice I expect Avenger to promote slower.

**Comment by xezon** [comment]
commy2:
> The Humvee is not mistakenly using a patriot missile. Although the model the Tow missiles uses is a patriot, but only when aiming at air.

> It is not clear to me that this change would enable Bombardement for the Avenger. It may already get Bombardment. Someone has to actually test these things.

**Comment by Float1ngFree** [comment]
> Although the model the Tow missiles uses is a patriot, but only when aiming at air.

I was talking about the air, this whole topic is about the air... c'mon guys!

It is not only the "model", as the projectile, as you know, specifies a lot of other critical params, such as `FuelLifetime` and `locomotor`:
![image](https://github.com/TheSuperHackers/GeneralsGamePatch/assets/17320562/8cd9f2d8-8571-433e-93f5-a59e5d2ca934)

That's why Humvee AA missile is so fast and basically never runs out of fuel, which guarantees the hit, unless there's PDL on jet.

> It is not clear to me that this change would enable Bombardement for the Avenger. It may already get Bombardment. Someone has to actually test these things.

`WeaponBonusPassedToPassengers` does enable the Bombardment bonus for avenger's turret. I did test it and same jet dies noticeably faster than w/o battle plan. I need to get fraps or something to record stuff & "show, don't tell".

**Comment by xezon** [comment]
> WeaponBonusPassedToPassengers does enable the Bombardment bonus for avenger's turret.

And the Bombardment bonus is not applied to Avenger weapon in retail game, correct?

**Comment by Float1ngFree** [comment]
> And the Bombardment bonus is not applied to Avenger weapon in retail game, correct?

Yes, because there's no `WeaponBonusPassedToPassengers` in Avenger's `OverlordContain` module. Adding it enables all weapon bonuses that apply to the Vehicle getting applied to portable_structure of Avenger Turret, which in turn hosts 2 anti-air lasers.

Currently, only "Hold the Line" applies to Avengers as it affects vehicle armor.

I guess that the idea behind Avengers was highest initial dps of all AA + missile defense, but at the cost of some range (300 vs 350) and lack of scaling through vet/weapon bonuses. Or they just forgot about `WeaponBonusPassedToPassengers` at all, which is typical in this mess.

Here, with `WeaponBonusPassedToPassengers = Yes` second raptor dies faster after Bombardment has been initialized.

https://github.com/TheSuperHackers/GeneralsGamePatch/assets/17320562/cc4e0959-e57e-4406-8350-9d1759e2cb5d

**Comment by Float1ngFree** [comment]
BTW the video is Raptor with countermeasures (80% in my patch) and Avenger with both AA laser dmg reduced from 10 to 9.

So in retail, a King Raptor or MiG that don't get the Countermeasures ArmorUpgrade - take 120% damage and with 10 avenger base damage they will melt even faster:
100 base dps +20% from Bombardment = 120 dps
200 MiG HP with armor upgrade, 120% damage taken = 200/1.2 = 166.6 dmg required. 166/120 = 1.38s to kill.

Compared to second raptor on video:
90 base dps +20% Bombardment = 108 dps
160 Raptor HP, 80% damage taken = 160/0.8 =200 dmg required, 200/108 = 1.85s to kill.

Now consider damage and rate of fire bonus from vet and it easily gets under a second.

So I'm not sure you really want to adopt this in 104p.

**Comment by xezon** [comment]
It is also possible to give lower bonuses. The global bonuses are not strict.

**Comment by Float1ngFree** [comment]
> It is also possible to give lower bonuses. The global bonuses are not strict.

I know as I've tweaked almost all of them in `GameData.ini`.
But they apply to all equally or they can be tweaked for specific unit/type?

**Comment by xezon** [comment]
> But they apply to all equally or they can be tweaked for specific unit/type?

Like so: https://github.com/TheSuperHackers/GeneralsGamePatch/pull/1055/files
